src/main/java/org/truffleruby/language/TruffleBootNodes.java

```diff
@@ -92,23 +91,23 @@ public DynamicObject forceContext() {
     @CoreMethod(names = "main", onSingleton = true)
     public abstract static class MainNode extends CoreMethodArrayArgumentsNode {
 
-        @Child private SnippetNode findSFileSnippetNode;
-        @Child private SnippetNode checkSyntaxSnippetNode;
-
         @Specialization
-        public int main(VirtualFrame frame, @Cached("create()") IndirectCallNode callNode) {
+        public int main(VirtualFrame frame,
+                @Cached("create()") IndirectCallNode callNode,
+                @Cached("new()") SnippetNode findSFileSnippetNode,
+                @Cached("new()") SnippetNode checkSyntaxSnippetNode) {
 
             setArgvGlobals();
 
-            Source source = loadMainSourceSettingDollarZero(frame);
+            Source source = loadMainSourceSettingDollarZero(frame, findSFileSnippetNode);
 
             if (source == null) {
                 // EXECUTION_ACTION was set to NONE
                 return 0;
             }
 
             if (getContext().getOptions().SYNTAX_CHECK) {
-                return (int) getCheckSyntaxSnippetNode().execute(
+                return (int) checkSyntaxSnippetNode.execute(
                         frame,
                         "Truffle::Boot.check_syntax source",
                         "source", source);
```

Comment:
FWIW I prefer a CallDispatchHeadNode rather than a Snippet when it's just one call.
It also does not need the frame if it's created with `CallDispatchHeadNode.createOnSelf()`
as then it does not need to check visibility.